File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3112_minimum-time-to-visit-disappearing-nodes/solutions/python3/solution.py
from typing import List
import heapq

class Solution:
    def minimumTime(self, n: int, edges: List[List[int]], disappear: List[int]) -> List[int]:
        # Build the graph as an adjacency list with edges being tuples (node, weight)
        graph = [[] for _ in range(n)]
        for u, v, length in edges:
            graph[u].append((v, length))
            graph[v].append((u, length))

        # Initialize the distance list with infinite distances, and set the start node 0 to 0
        dist = [float('inf')] * n
        dist[0] = 0

        # Priority queue for processing nodes based on shortest discovered distance
        pq = [(0, 0)]  # (distance, node)

        # Dijkstra's algorithm
        while pq:
            d, node = heapq.heappop(pq)

            # If current distance is greater than the distance in the array, skip it
            if d > dist[node]:
                continue

            # Explore neighbors
            for neighbor, length in graph[node]:
                new_dist = d + length
                if new_dist < dist[neighbor]:
                    dist[neighbor] = new_dist
                    heapq.heappush(pq, (new_dist, neighbor))

        # Compare the minimum distance with disappear times
        answer = []
        for i in range(n):
            if dist[i] <= disappear[i]:
                answer.append(dist[i])
            else:
                answer.append(-1)

        return answer

File: DataSets/DataSet-New/LeetCodeTasks - Develop/LeetCodeTasks/3112_minimum-time-to-visit-disappearing-nodes/solutions/java/solution.java
import java.util.*;

class Solution {
    public int[] minimumTime(int n, int[][] edges, int[] disappear) {
        // Graph adjacency list
        List<List<int[]>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }
        for (int[] edge : edges) {
            int u = edge[0], v = edge[1], w = edge[2];
            graph.get(u).add(new int[] {v, w});
            graph.get(v).add(new int[] {u, w});
        }
        
        // Dijkstra's algorithm initialization
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[1]));
        pq.offer(new int[] {0, 0});  // start node, distance
        int[] dist = new int[n];
        Arrays.fill(dist, Integer.MAX_VALUE);
        dist[0] = 0;
        
        // Process nodes
        while (!pq.isEmpty()) {
            int[] cur = pq.poll();
            int node = cur[0];
            int d = cur[1];
            
            if (d > dist[node]) continue;
            
            for (int[] neighbor : graph.get(node)) {
                int nextNode = neighbor[0];
                int weight = neighbor[1];
                int nextDist = d + weight;
                if (nextDist < dist[nextNode]) {
                    dist[nextNode] = nextDist;
                    pq.offer(new int[] {nextNode, nextDist});
                }
            }
        }
        
        // Final result checking disappearance times
        int[] result = new int[n];
        for (int i = 0; i < n; i++) {
            if (dist[i] <= disappear[i] - 1) {
                result[i] = dist[i];
            } else {
                result[i] = -1;
            }
        }
        return result;
    }
}
